dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp

```diff
@@ -363,22 +386,27 @@ void LocalAdmissionController::watchGAC()
         // 2. grpc stream read error.
         // 3. watch is cancel.
         // Will sleep and try again.
-        std::unique_lock lock(mu);
-        if (cv.wait_for(lock, std::chrono::seconds(10), [this]() { return stopped.load(); }))
-            return;
+        {
+            std::unique_lock lock(mu);
+            if (cv.wait_for(lock, std::chrono::seconds(10), [this]() { return stopped.load(); }))
+                return;
+
+            // Create new grpc_context for each reader/writer.
+            watch_gac_grpc_context = std::make_unique<grpc::ClientContext>();
+        }
     }
 }
 
 void LocalAdmissionController::doWatch()
 {
-    auto stream = etcd_client->watch(&watch_gac_grpc_context);
+    auto stream = etcd_client->watch(watch_gac_grpc_context.get());
```

Comment:
Looks like a lock is needed here
```
grpc::ClientContext * grpc_ctx;
{
    std::lock_guard lock(mu);
    grpc_ctx = watch_gac_grpc_context.get();
}
RUNTIME_ASSERT(grpc_ctx != nullptr);
```

Reply:
Or using `Common/PtrHolder`

Reply:
no need to lock here?
because write and read of `watch_gac_grpc_context` are all in watch_gac thread.

Reply:
`TryCancel` is called in another thread, but it's thread safe

Reply:
ok